reactioncommerce/reaction
# Remove aldeed:schema-index Meteor package, use collectionIndex

Resolves #5080   
Impact: **minor**  
Type: **feature**

## Changes
The `aldeed:schema-index` Meteor package takes SimpleSchema options like `index: 1` and `unique: true` and calls MongoDB `createIndex` based on them. In an effort to reduce Meteor dependencies, all usage of this and the `aldeed:schema-index` dependency itself has now been removed.

Indexes are now set up by the plugins that care about them. They call `collectionIndex` in a startup function to do this. This may not be the final way this is done, but this PR is one step out of many to get to where we are going.

The goal was not to make any decisions about the appropriateness of any indexes in this PR. I did notice some that seem to be outdated, but I did not remove them at this time. This is strictly an internal refactor.

**There is one way in which this PR is more than a refactor. There were several fields in the Catalog schema that were marked to be indexed, but because that schema was not attached to the Catalog collection in Meteor, the indexes were not actually created. I did copy the useful ones to the new code, so now Catalog collection will have additional indexes created.**

## Breaking changes
If you have custom plugins that use `index` or `unique` options in their schemas, add back the `aldeed:schema-index` dependency or update the plugins to use `collectionIndex` function.

## Testing
1. Run the app on `develop` branch. Verify that your local database has indexes. Save off the list of indexes. For example:

```
db.getCollectionNames().forEach(function(collection) {
   indexes = db[collection].getIndexes();
   print("Indexes for " + collection + ":");
   printjson(indexes);
});
```

2. Drop all collections.
3. Run the app on this PR branch. Verify that your local database has indexes. Compare the list with the one you saved off. It should have all of the same indexes as before, and perhaps a few additional. (As stated above, some indexes may not make sense anymore, but we will evaluate this in a future PR.)
